File: packages/core-ts/src/routing/memo.ts
import { Warning } from "../address/types";
import { KnownMemoType } from "./types";

export type NormalizeResult = {
  normalized: string | null;
  warnings: Warning[];
};

export type MemoValidationResult = {
  valid: boolean;
  type: string;
  normalizedValue: string | null;
  warnings: Warning[];
  error?: string;
};

export type GeneratedMemo = {
  type: KnownMemoType;
  value: string;
};

/**
 * Maximum value for a 64-bit unsigned integer (uint64).
 */
const UINT64_MAX = BigInt("18446744073709551615");

/**
 * Calculates UTF-8 byte length of a string without external dependencies.
 */
function getUtf8ByteLength(str: string): number {
  if (typeof TextEncoder !== "undefined") {
    return new TextEncoder().encode(str).length;
  }
  return Buffer.byteLength(str, "utf8");
}

/**
 * Normalizes a numeric string into a canonical uint64 representation.
 * Strips leading zeros and validates that the value is within uint64 boundaries.
 * 
 * @param s - The numeric string to normalize.
 * @returns Result containing the normalized string (or null if invalid) and any warnings.
 */
export function normalizeMemoTextId(s: string): NormalizeResult {
  const warnings: Warning[] = [];

  if (s.length === 0 || !/^\d+$/.test(s)) {
    return { normalized: null, warnings };
  }

  let normalized = s.replace(/^0+/, "");
  if (normalized === "") {
    normalized = "0";
  }

  if (normalized !== s) {
    warnings.push({
      code: "NON_CANONICAL_ROUTING_ID",
      severity: "warn",
      message:
        "Memo routing ID had leading zeros. Normalized to canonical decimal.",
      normalization: { original: s, normalized },
    });
  }

  try {
    const val = BigInt(normalized);
    if (val > UINT64_MAX) {
      return { normalized: null, warnings };
    }
  } catch {
    return { normalized: null, warnings };
  }

  return { normalized, warnings };
}

/**
 * Validates a MEMO_TEXT string (must be <= 28 UTF-8 bytes).
 */
export function validateMemoText(text: string): boolean {
  if (typeof text !== "string") return false;
  const len = getUtf8ByteLength(text);
  return len > 0 && len <= 28;
}

/**
 * Validates a MEMO_HASH string (must be 64-character hexadecimal representation of 32 bytes).
 */
export function validateMemoHash(hash: string): boolean {
  if (typeof hash !== "string") return false;
  return /^[0-9a-fA-F]{64}$/.test(hash);
}

/**
 * Validates a MEMO_ID string or bigint (must be a valid uint64: 0 to 18446744073709551615).
 */
export function validateMemoId(id: string | bigint): boolean {
  if (typeof id === "bigint") {
    return id >= BigInt(0) && id <= UINT64_MAX;
  }
  if (typeof id !== "string") return false;
  const norm = normalizeMemoTextId(id);
  return norm.normalized !== null;
}

/**
 * Unified validator for Stellar memos (none, id, text, hash, return).
 */
export function validateMemo(
  type: KnownMemoType | string,
  value: string | null | undefined
): MemoValidationResult {
  const warnings: Warning[] = [];

  if (type === "none") {
    return {
      valid: value === null || value === undefined || value === "",
      type: "none",
      normalizedValue: null,
      warnings,
      error: value ? "Memo value provided for memo type 'none'" : undefined,
    };
  }

  if (!value) {
    return {
      valid: false,
      type,
      normalizedValue: null,
      warnings,
      error: `Memo value is required for memo type '${type}'`,
    };
  }

  switch (type) {
    case "id": {
      const norm = normalizeMemoTextId(value);
      warnings.push(...norm.warnings);
      if (norm.normalized === null) {
        return {
          valid: false,
          type: "id",
          normalizedValue: null,
          warnings,
          error: "MEMO_ID must be a numeric integer between 0 and 18446744073709551615",
        };
      }
      return {
        valid: true,
        type: "id",
        normalizedValue: norm.normalized,
        warnings,
      };
    }

    case "text": {
      if (!validateMemoText(value)) {
        return {
          valid: false,
          type: "text",
          normalizedValue: null,
          warnings,
          error: "MEMO_TEXT must not be empty and must be at most 28 UTF-8 bytes",
        };
      }
      return {
        valid: true,
        type: "text",
        normalizedValue: value,
        warnings,
      };
    }

    case "hash":
    case "return": {
      if (!validateMemoHash(value)) {
        return {
          valid: false,
          type,
          normalizedValue: null,
          warnings,
          error: `MEMO_${type.toUpperCase()} must be a 64-character hex string (32 bytes)`,
        };
      }
      return {
        valid: true,
        type,
        normalizedValue: value.toLowerCase(),
        warnings,
      };
    }

    default:
      return {
        valid: false,
        type,
        normalizedValue: null,
        warnings,
        error: `Unrecognized memo type: '${type}'`,
      };
  }
}

/**
 * Generates a valid MEMO_TEXT value (ensures max 28 UTF-8 bytes).
 */
export function generateMemoText(text: string): string {
  if (!text || typeof text !== "string") {
    throw new Error("Input text must be a non-empty string");
  }
  let result = text;
  while (getUtf8ByteLength(result) > 28) {
    result = result.slice(0, -1);
  }
  if (!result) {
    throw new Error("Failed to generate valid MEMO_TEXT");
  }
  return result;
}

/**
 * Generates a valid MEMO_HASH 64-char lowercase hex string.
 * Accepts a 64-char hex string, a 32-byte Uint8Array, or pads/formats string.
 */
export function generateMemoHash(input: string | Uint8Array): string {
  if (input instanceof Uint8Array) {
    if (input.length !== 32) {
      throw new Error("Uint8Array input for MEMO_HASH must be exactly 32 bytes");
    }
    return Array.from(input)
      .map((b) => b.toString(16).padStart(2, "0"))
      .join("");
  }

  if (typeof input === "string") {
    const trimmed = input.trim();
    if (/^[0-9a-fA-F]{64}$/.test(trimmed)) {
      return trimmed.toLowerCase();
    }
    // Hex string shorter than 64 chars -> pad with leading zeros
    if (/^[0-9a-fA-F]+$/.test(trimmed) && trimmed.length <= 64) {
      return trimmed.padStart(64, "0").toLowerCase();
    }
  }

  throw new Error("Invalid input for MEMO_HASH: expected 64-char hex string or 32-byte Uint8Array");
}

/**
 * Generates a valid canonical MEMO_ID string.
 */
export function generateMemoId(id: string | bigint): string {
  if (typeof id === "bigint") {
    if (id < BigInt(0) || id > UINT64_MAX) {
      throw new Error("MEMO_ID bigint out of uint64 bounds");
    }
    return id.toString();
  }
  const norm = normalizeMemoTextId(id);
  if (!norm.normalized) {
    throw new Error("Invalid MEMO_ID: must be a valid uint64 numeric string");
  }
  return norm.normalized;
}

/**
 * Unified memo generator for Stellar transactions.
 */
export function generateMemo(
  type: KnownMemoType,
  value: string | bigint | Uint8Array
): GeneratedMemo {
  switch (type) {
    case "text":
      if (typeof value !== "string") {
        throw new Error("MEMO_TEXT requires a string value");
      }
      return { type: "text", value: generateMemoText(value) };
    case "hash":
      if (typeof value !== "string" && !(value instanceof Uint8Array)) {
        throw new Error("MEMO_HASH requires a hex string or Uint8Array value");
      }
      return { type: "hash", value: generateMemoHash(value) };
    case "id":
      if (typeof value !== "string" && typeof value !== "bigint") {
        throw new Error("MEMO_ID requires a string or bigint value");
      }
      return { type: "id", value: generateMemoId(value) };
    default:
      throw new Error(`Unsupported memo generation type: '${type}'`);
  }
}

File: packages/core-go/routing/memo.go
package routing

import (
	"encoding/hex"
	"errors"
	"fmt"
	"math/big"
	"regexp"
	"strings"

	"github.com/Boxkit-Labs/stellar-address-kit/packages/core-go/address"
)

var digitsOnly = regexp.MustCompile(`^\d+$`)
var hex64Regex = regexp.MustCompile(`^[0-9a-fA-F]{64}$`)
var uint64Max, _ = new(big.Int).SetString("18446744073709551615", 10)

type NormalizeResult struct {
	Normalized string
	Warnings   []address.Warning
}

type MemoValidationResult struct {
	Valid           bool
	Type            string
	NormalizedValue string
	Warnings        []address.Warning
	Error           string
}

type GeneratedMemo struct {
	Type  string
	Value string
}

func NormalizeMemoTextID(s string) NormalizeResult {
	warnings := []address.Warning{}

	if s == "" || !digitsOnly.MatchString(s) {
		return NormalizeResult{Normalized: "", Warnings: warnings}
	}

	normalized := strings.TrimLeft(s, "0")
	if normalized == "" {
		normalized = "0"
	}

	if normalized != s {
		warnings = append(warnings, address.Warning{
			Code:     address.WarnNonCanonicalRoutingID,
			Severity: "warn",
			Message:  "Memo routing ID had leading zeros. Normalized to canonical decimal.",
			Normalization: &address.Normalization{
				Original:   s,
				Normalized: normalized,
			},
		})
	}

	val, ok := new(big.Int).SetString(normalized, 10)
	if !ok || val.Cmp(uint64Max) > 0 {
		return NormalizeResult{Normalized: "", Warnings: warnings}
	}

	return NormalizeResult{Normalized: normalized, Warnings: warnings}
}

// ValidateMemoText validates a MEMO_TEXT string (must be non-empty and <= 28 UTF-8 bytes).
func ValidateMemoText(text string) bool {
	length := len([]byte(text))
	return length > 0 && length <= 28
}

// ValidateMemoHash validates a MEMO_HASH string (must be 64-char hex representing 32 bytes).
func ValidateMemoHash(hash string) bool {
	return hex64Regex.MatchString(hash)
}

// ValidateMemoID validates a MEMO_ID string (must be uint64).
func ValidateMemoID(id string) bool {
	norm := NormalizeMemoTextID(id)
	return norm.Normalized != ""
}

// ValidateMemo is a unified validator for Stellar memos.
func ValidateMemo(memoType, value string) MemoValidationResult {
	warnings := []address.Warning{}

	if memoType == "none" {
		valid := value == ""
		errStr := ""
		if !valid {
			errStr = "Memo value provided for memo type 'none'"
		}
		return MemoValidationResult{
			Valid:           valid,
			Type:            "none",
			NormalizedValue: "",
			Warnings:        warnings,
			Error:           errStr,
		}
	}

	if value == "" {
		return MemoValidationResult{
			Valid:           false,
			Type:            memoType,
			NormalizedValue: "",
			Warnings:        warnings,
			Error:           fmt.Sprintf("Memo value is required for memo type '%s'", memoType),
		}
	}

	switch memoType {
	case "id":
		norm := NormalizeMemoTextID(value)
		warnings = append(warnings, norm.Warnings...)
		if norm.Normalized == "" {
			return MemoValidationResult{
				Valid:           false,
				Type:            "id",
				NormalizedValue: "",
				Warnings:        warnings,
				Error:           "MEMO_ID must be a numeric integer between 0 and 18446744073709551615",
			}
		}
		return MemoValidationResult{
			Valid:           true,
			Type:            "id",
			NormalizedValue: norm.Normalized,
			Warnings:        warnings,
		}

	case "text":
		if !ValidateMemoText(value) {
			return MemoValidationResult{
				Valid:           false,
				Type:            "text",
				NormalizedValue: "",
				Warnings:        warnings,
				Error:           "MEMO_TEXT must not be empty and must be at most 28 UTF-8 bytes",
			}
		}
		return MemoValidationResult{
			Valid:           true,
			Type:            "text",
			NormalizedValue: value,
			Warnings:        warnings,
		}

	case "hash", "return":
		if !ValidateMemoHash(value) {
			return MemoValidationResult{
				Valid:           false,
				Type:            memoType,
				NormalizedValue: "",
				Warnings:        warnings,
				Error:           fmt.Sprintf("MEMO_%s must be a 64-character hex string (32 bytes)", strings.ToUpper(memoType)),
			}
		}
		return MemoValidationResult{
			Valid:           true,
			Type:            memoType,
			NormalizedValue: strings.ToLower(value),
			Warnings:        warnings,
		}

	default:
		return MemoValidationResult{
			Valid:           false,
			Type:            memoType,
			NormalizedValue: "",
			Warnings:        warnings,
			Error:           fmt.Sprintf("Unrecognized memo type: '%s'", memoType),
		}
	}
}

// GenerateMemoText formats string to valid 28-byte MEMO_TEXT.
func GenerateMemoText(text string) (string, error) {
	if text == "" {
		return "", errors.New("input text must be a non-empty string")
	}
	bytes := []byte(text)
	if len(bytes) <= 28 {
		return text, nil
	}
	// Truncate to 28 bytes cleanly
	return string(bytes[:28]), nil
}

// GenerateMemoHash formats or pads input string to a 64-char hex string.
func GenerateMemoHash(input string) (string, error) {
	trimmed := strings.TrimSpace(input)
	if hex64Regex.MatchString(trimmed) {
		return strings.ToLower(trimmed), nil
	}
	if len(trimmed) <= 64 {
		_, err := hex.DecodeString(trimmed)
		if err == nil {
			padded := fmt.Sprintf("%064s", trimmed)
			return strings.ToLower(padded), nil
		}
	}
	return "", errors.New("invalid input for MEMO_HASH: expected hex string up to 64 chars")
}

// GenerateMemoID normalizes id string into canonical uint64 decimal string.
func GenerateMemoID(id string) (string, error) {
	norm := NormalizeMemoTextID(id)
	if norm.Normalized == "" {
		return "", errors.New("invalid MEMO_ID: must be a valid uint64 numeric string")
	}
	return norm.Normalized, nil
}

// GenerateMemo produces a unified GeneratedMemo struct.
func GenerateMemo(memoType, value string) (GeneratedMemo, error) {
	switch memoType {
	case "text":
		val, err := GenerateMemoText(value)
		if err != nil {
			return GeneratedMemo{}, err
		}
		return GeneratedMemo{Type: "text", Value: val}, nil
	case "hash":
		val, err := GenerateMemoHash(value)
		if err != nil {
			return GeneratedMemo{}, err
		}
		return GeneratedMemo{Type: "hash", Value: val}, nil
	case "id":
		val, err := GenerateMemoID(value)
		if err != nil {
			return GeneratedMemo{}, err
		}
		return GeneratedMemo{Type: "id", Value: val}, nil
	default:
		return GeneratedMemo{}, fmt.Errorf("unsupported memo generation type: '%s'", memoType)
	}
}
